File: example_py/seer_audio_example.py
#!/usr/bin/env python3

import argparse
import logging
import signal
import time
from pathlib import Path

import flexiv_amr


logger = logging.getLogger("seer_audio_example")
stop_requested = False
UPLOAD_READY_DELAY_SECONDS = 1.0
STATUS_POLL_INTERVAL_SECONDS = 0.3
STATUS_POLL_COUNT = 8


def signal_handler(signum, frame) -> None:
    global stop_requested
    stop_requested = True
    logger.info("Stop signal received.")


def is_audio_file_path(path: str) -> bool:
    return Path(path).suffix.lower() == ".wav"


def audio_name_from_path(path: str) -> str:
    name = Path(path).stem
    if not name:
        raise ValueError("audio file path must contain a file name")
    return name


def parse_arguments() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Upload, download, play, or pause audio on a SEER AMR."
    )
    parser.add_argument("amr_ip", help="SEER AMR IPv4 address or hostname")
    parser.add_argument(
        "audio",
        nargs="?",
        help="Audio name stored on the AMR, or a local .wav file to upload first",
    )
    parser.add_argument(
        "--loop",
        action="store_true",
        help="Loop playback when playing audio",
    )
    parser.add_argument(
        "--download",
        metavar="OUTPUT_FILE",
        help="Download the named audio to this local file path",
    )
    parser.add_argument(
        "--pause",
        action="store_true",
        help="Pause current audio playback",
    )
    args = parser.parse_args()

    if not args.pause and not args.audio:
        parser.error("audio is required unless --pause is used alone")
    if args.download and not args.audio:
        parser.error("audio is required when downloading audio")
    if args.audio and Path(args.audio).suffix.lower() == ".mp3":
        parser.error("only .wav audio files are supported")

    return args


def log_audio_status(amr: flexiv_amr.SeerAmrClient, context: str):
    status = amr.get_audio_status()
    logger.info(
        "%s audio status: status=%s, sound_name='%s', loop=%s, count=%s",
        context,
        status.status,
        status.sound_name,
        status.loop,
        status.count,
    )
    return status


def verify_uploaded_audio(
    amr: flexiv_amr.SeerAmrClient, audio_name: str, upload_file: str
) -> None:
    logger.info(
        "Waiting %.0f ms for the AMR to index uploaded audio ...",
        UPLOAD_READY_DELAY_SECONDS * 1000.0,
    )
    time.sleep(UPLOAD_READY_DELAY_SECONDS)

    local_data = Path(upload_file).read_bytes()
    logger.info("Verifying upload by downloading audio '%s' ...", audio_name)
    downloaded_data = amr.download_audio(audio_name)

    if len(downloaded_data) != len(local_data):
        logger.warning(
            "Downloaded size %d differs from local size %d. The upload was readable "
            "by name, but the bytes are not identical.",
            len(downloaded_data),
            len(local_data),
        )
        return

    if downloaded_data != local_data:
        logger.warning(
            "Downloaded audio '%s' has the same size as the local file but different content.",
            audio_name,
        )
        return

    logger.info(
        "Upload verification succeeded: '%s' is readable by name and bytes match.",
        audio_name,
    )


def poll_audio_status_after_play(amr: flexiv_amr.SeerAmrClient, audio_name: str) -> None:
    accepted_names = {audio_name, f"{audio_name}.wav"}
    for _ in range(STATUS_POLL_COUNT):
        time.sleep(STATUS_POLL_INTERVAL_SECONDS)
        status = log_audio_status(amr, "After play")
        if status.sound_name in accepted_names:
            logger.info("AMR reports requested audio '%s' as current sound.", audio_name)
            return

    logger.warning(
        "AMR did not report '%s' as current sound after the play command. If upload "
        "verification succeeded, check audio format/codec support and robot speaker configuration.",
        audio_name,
    )


def main() -> int:
    logging.basicConfig(
        level=logging.INFO,
        format="[%(asctime)s.%(msecs)03d] [%(levelname)s] %(message)s",
        datefmt="%Y-%m-%d %H:%M:%S",
    )
    args = parse_arguments()

    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)

    options = flexiv_amr.SeerAmrOptions()
    options.host = args.amr_ip
    options.auto_connect = False
    options.auto_reconnect = True

    amr = flexiv_amr.SeerAmrClient(options)
    connected = False

    try:
        logger.info("Connecting to SEER AMR: %s ...", args.amr_ip)
        amr.connect()
        connected = True
        logger.info("Connected: %s", amr.is_connected())

        log_audio_status(amr, "Initial")

        if args.pause:
            logger.info("Pausing audio playback ...")
            if not amr.pause_audio():
                raise RuntimeError("SEER AMR rejected the pause audio command")
            logger.info("Audio pause command accepted.")
            time.sleep(0.2)
            log_audio_status(amr, "After pause")

        audio_name = args.audio or ""
        upload_file = ""
        if audio_name and is_audio_file_path(audio_name):
            upload_file = audio_name
            audio_name = audio_name_from_path(upload_file)

        if args.download:
            logger.info("Downloading audio '%s' to '%s' ...", audio_name, args.download)
            audio_data = amr.download_audio(audio_name)
            Path(args.download).write_bytes(audio_data)
            logger.info("Downloaded %d bytes.", len(audio_data))

        if upload_file:
            logger.info("Uploading audio '%s' from '%s' ...", audio_name, upload_file)
            if not amr.upload_audio(upload_file):
                raise RuntimeError("SEER AMR rejected the upload audio command")
            logger.info("Audio upload accepted.")
            verify_uploaded_audio(amr, audio_name, upload_file)

        if audio_name and not args.download and not stop_requested:
            logger.info("Playing audio '%s', loop=%s ...", audio_name, args.loop)
            if not amr.play_audio(audio_name, args.loop):
                raise RuntimeError("SEER AMR rejected the play audio command")
            logger.info("Audio play command accepted.")
            poll_audio_status_after_play(amr, audio_name)

        logger.info("Disconnecting...")
        amr.disconnect()
        connected = False
        logger.info("Audio example completed.")
        return 0

    except flexiv_amr.AmrException as e:
        logger.error("AMR exception: %s", e)

    except Exception as e:
        logger.exception("Unexpected exception: %s", e)

    finally:
        if connected:
            try:
                amr.disconnect()
            except Exception:
                pass

    return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: example/seer_audio_example.cpp
/**
 * @example seer_audio_example.cpp
 * This example uploads, downloads, plays, or pauses audio on a SEER AMR.
 */

#include <algorithm>
#include <chrono>
#include <cctype>
#include <cstdint>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <iterator>
#include <stdexcept>
#include <string>
#include <thread>
#include <vector>

#include <flexiv/amr/exceptions.h>
#include <flexiv/amr/vendor/seer/seer_amr.h>
#include <spdlog/spdlog.h>

namespace {

struct ProgramOptions {
    std::string amr_ip;
    std::string audio_name;
    std::string upload_file;
    std::string download_file;
    bool loop = false;
    bool pause = false;
};

constexpr auto kUploadReadyDelay = std::chrono::seconds(1);
constexpr auto kStatusPollInterval = std::chrono::milliseconds(300);
constexpr int kStatusPollCount = 8;

void PrintUsage(const char* program) {
    std::cerr
        << "Usage:\n"
        << "  " << program
        << " <amr_ip> [audio_name|audio_file] [--loop] [--download output_file] [--pause]\n\n"
        << "Arguments:\n"
        << "  audio_name  Audio name stored on the AMR, without file extension.\n"
        << "  audio_file  Local .wav file. If provided, upload it before playback.\n\n"
        << "Examples:\n"
        << "  " << program << " 192.168.192.5 welcome --loop\n"
        << "  " << program << " 192.168.192.5 ./welcome.wav\n"
        << "  " << program << " 192.168.192.5 welcome --download ./welcome.wav\n"
        << "  " << program << " 192.168.192.5 --pause\n";
}

std::string ToLower(std::string text) {
    std::transform(text.begin(), text.end(), text.begin(), [](unsigned char ch) {
        return static_cast<char>(std::tolower(ch));
    });
    return text;
}

bool IsAudioFilePath(const std::string& path) {
    const auto extension = ToLower(std::filesystem::path(path).extension().string());
    return extension == ".wav";
}

bool HasUnsupportedAudioExtension(const std::string& path) {
    const auto extension = ToLower(std::filesystem::path(path).extension().string());
    return extension == ".mp3";
}

std::string AudioNameFromPath(const std::string& path) {
    const auto stem = std::filesystem::path(path).stem().string();
    if (stem.empty()) {
        throw std::invalid_argument("audio file path must contain a file name");
    }
    return stem;
}

ProgramOptions ParseArguments(int argc, char* argv[]) {
    if (argc < 2) {
        throw std::invalid_argument("missing required arguments");
    }

    ProgramOptions options;
    options.amr_ip = argv[1];

    for (int i = 2; i < argc; ++i) {
        const std::string arg = argv[i];
        if (arg == "--loop") {
            options.loop = true;
        } else if (arg == "--pause") {
            options.pause = true;
        } else if (arg == "--download") {
            if (i + 1 >= argc) {
                throw std::invalid_argument("--download requires an output file path");
            }
            options.download_file = argv[++i];
        } else if (HasUnsupportedAudioExtension(arg)) {
            throw std::invalid_argument("only .wav audio files are supported");
        } else if (IsAudioFilePath(arg) && options.upload_file.empty()) {
            options.upload_file = arg;
        } else if (options.audio_name.empty()) {
            options.audio_name = arg;
        } else {
            throw std::invalid_argument("unexpected argument: " + arg);
        }
    }

    if (!options.upload_file.empty()) {
        options.audio_name = AudioNameFromPath(options.upload_file);
    }
    if (!options.pause && options.audio_name.empty()) {
        throw std::invalid_argument("audio_name is required unless --pause is used alone");
    }
    if (!options.download_file.empty() && options.audio_name.empty()) {
        throw std::invalid_argument("audio_name is required when downloading audio");
    }

    return options;
}

void WriteBinaryFile(const std::string& path, const std::vector<uint8_t>& data) {
    std::ofstream file(path, std::ios::binary);
    if (!file) {
        throw std::runtime_error("failed to open output file: " + path);
    }

    file.write(reinterpret_cast<const char*>(data.data()),
               static_cast<std::streamsize>(data.size()));
    if (!file) {
        throw std::runtime_error("failed to write output file: " + path);
    }
}

std::vector<uint8_t> ReadBinaryFile(const std::string& path) {
    std::ifstream file(path, std::ios::binary);
    if (!file) {
        throw std::runtime_error("failed to open input file: " + path);
    }

    return std::vector<uint8_t>((std::istreambuf_iterator<char>(file)),
                                std::istreambuf_iterator<char>());
}

flexiv::amr::seer::SeerAudioStatus LogAudioStatus(flexiv::amr::seer::SeerAmrClient& amr,
                                                  const char* context) {
    const auto status = amr.GetAudioStatus();
    spdlog::info("{} audio status: status={}, sound_name='{}', loop={}, count={}",
                 context,
                 status.status,
                 status.sound_name,
                 status.loop,
                 status.count);
    return status;
}

void VerifyUploadedAudio(flexiv::amr::seer::SeerAmrClient& amr,
                         const std::string& audio_name,
                         const std::string& upload_file) {
    spdlog::info("Waiting {} ms for the AMR to index uploaded audio ...",
                 std::chrono::duration_cast<std::chrono::milliseconds>(kUploadReadyDelay).count());
    std::this_thread::sleep_for(kUploadReadyDelay);

    const auto local_data = ReadBinaryFile(upload_file);
    spdlog::info("Verifying upload by downloading audio '{}' ...", audio_name);
    const auto downloaded_data = amr.DownloadAudio(audio_name);

    if (downloaded_data.size() != local_data.size()) {
        spdlog::warn("Downloaded size {} differs from local size {}. The upload was readable by "
                     "name, but the bytes are not identical.",
                     downloaded_data.size(),
                     local_data.size());
        return;
    }

    if (downloaded_data != local_data) {
        spdlog::warn("Downloaded audio '{}' has the same size as the local file but different "
                     "content.",
                     audio_name);
        return;
    }

    spdlog::info("Upload verification succeeded: '{}' is readable by name and bytes match.",
                 audio_name);
}

void PollAudioStatusAfterPlay(flexiv::amr::seer::SeerAmrClient& amr,
                              const std::string& audio_name) {
    for (int i = 0; i < kStatusPollCount; ++i) {
        std::this_thread::sleep_for(kStatusPollInterval);
        const auto status = LogAudioStatus(amr, "After play");
        if (status.sound_name == audio_name || status.sound_name == audio_name + ".wav") {
            spdlog::info("AMR reports requested audio '{}' as current sound.", audio_name);
            return;
        }
    }

    spdlog::warn("AMR did not report '{}' as current sound after the play command. If upload "
                 "verification succeeded, check audio format/codec support and robot speaker "
                 "configuration.",
                 audio_name);
}

}  // namespace

int main(int argc, char* argv[]) {
    spdlog::set_pattern("[%Y-%m-%d %H:%M:%S.%e] [%^%l%$] %v");

    ProgramOptions program_options;
    try {
        program_options = ParseArguments(argc, argv);
    } catch (const std::exception& e) {
        spdlog::error("Invalid arguments: {}", e.what());
        PrintUsage(argv[0]);
        return 1;
    }

    try {
        flexiv::amr::seer::SeerAmrOptions options;
        options.host = program_options.amr_ip;
        options.auto_connect = false;

        flexiv::amr::seer::SeerAmrClient amr(options);

        spdlog::info("Connecting to SEER AMR: {} ...", options.host);
        amr.Connect();
        spdlog::info("Connected: {}", amr.IsConnected());
        LogAudioStatus(amr, "Initial");

        if (program_options.pause) {
            spdlog::info("Pausing audio playback ...");
            if (!amr.PauseAudio()) {
                throw std::runtime_error("SEER AMR rejected the pause audio command");
            }
            spdlog::info("Audio pause command accepted.");
            std::this_thread::sleep_for(std::chrono::milliseconds(200));
            LogAudioStatus(amr, "After pause");
        }

        if (!program_options.download_file.empty()) {
            spdlog::info("Downloading audio '{}' to '{}' ...",
                         program_options.audio_name,
                         program_options.download_file);

            const auto audio_data = amr.DownloadAudio(program_options.audio_name);
            WriteBinaryFile(program_options.download_file, audio_data);

            spdlog::info("Downloaded {} bytes.", audio_data.size());
        }

        if (!program_options.upload_file.empty()) {
            spdlog::info("Uploading audio '{}' from '{}' ...",
                         program_options.audio_name,
                         program_options.upload_file);

            if (!amr.UploadAudio(program_options.upload_file)) {
                throw std::runtime_error("SEER AMR rejected the upload audio command");
            }
            spdlog::info("Audio upload accepted.");
            VerifyUploadedAudio(amr, program_options.audio_name, program_options.upload_file);
        }

        if (!program_options.audio_name.empty() && program_options.download_file.empty()) {
            spdlog::info("Playing audio '{}', loop={} ...",
                         program_options.audio_name,
                         program_options.loop);
            if (!amr.PlayAudio(program_options.audio_name, program_options.loop)) {
                throw std::runtime_error("SEER AMR rejected the play audio command");
            }

            spdlog::info("Audio play command accepted.");
            PollAudioStatusAfterPlay(amr, program_options.audio_name);
        }

        amr.Disconnect();
        spdlog::info("Audio example completed.");
        return 0;
    } catch (const flexiv::amr::AmrException& e) {
        spdlog::error("AMR exception: {}, code: {}", e.what(), static_cast<int>(e.code()));
    } catch (const std::exception& e) {
        spdlog::error("std exception: {}", e.what());
    }

    return 1;
}
